File: python/minimize_the_maximum_difference_of_pairs.py
class Solution:
    def minimizeMax(self, nums: list[int], p: int) -> int:
        nums.sort()
        result: int = 0
        l: int = 0
        r: int = 10**9
        while l <= r:
            mid: int = (l + r) // 2
            if self.isValid(mid, nums, p):
                r = mid - 1
                result = mid
            else:
                l = mid + 1

        return result

    def isValid(self, diff: int, nums: list[int], p: int) -> bool:
        i: int = 0
        count: int = 0
        while i < len(nums) - 1:
            if abs(nums[i] - nums[i + 1]) <= diff:
                i += 2
                count += 1
            else:
                i += 1

        return count >= p


def test_minimizeMax_case_1():
    # arrange
    nums: list[int] = [10, 1, 2, 7, 1, 3]
    p: int = 2
    expected: int = 1

    # act
    solution = Solution()
    actual = solution.minimizeMax(nums, p)

    # assert
    assert expected == actual


def test_minimizeMax_case_2():
    # arrange
    nums: list[int] = [4, 2, 1, 2]
    p: int = 1
    expected: int = 0

    # act
    solution = Solution()
    actual = solution.minimizeMax(nums, p)

    # assert
    assert expected == actual

File: rust/src/code/minimize_the_maximum_difference_of_pairs.rs
struct Solution {}

impl Solution {
    // NOTE: time complexity: O(nlogn)
    pub fn minimize_max(mut nums: Vec<i32>, p: i32) -> i32 {
        nums.sort();
        let mut l: i32 = 0;
        let mut r: i32 = 10_i32.pow(9);
        let mut result: i32 = r;
        while l <= r {
            let mid = (l + r) / 2;
            if Self::is_valid(mid, &nums, p) {
                result = mid;
                if mid.checked_sub(1).is_none() {
                    break;
                }
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        result
    }

    fn is_valid(diff: i32, nums: &Vec<i32>, p: i32) -> bool {
        let mut i: usize = 0;
        let mut count: i32 = 0;
        while i < nums.len() - 1 {
            if nums[i + 1] - nums[i] <= diff {
                i += 2;
                count += 1;
            } else {
                i += 1;
            }
        }
        count >= p
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_minimize_max_case_1() {
        // arrange
        let nums: Vec<i32> = vec![10, 1, 2, 7, 1, 3];
        let p: i32 = 2;
        let expected: i32 = 1;

        // act
        let actual = Solution::minimize_max(nums, p);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_minimize_max_case_2() {
        // arrange
        let nums: Vec<i32> = vec![4, 2, 1, 2];
        let p: i32 = 1;
        let expected: i32 = 0;

        // act
        let actual = Solution::minimize_max(nums, p);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_minimize_max_case_3() {
        // arrange
        let nums: Vec<i32> = vec![0, 5, 3, 4];
        let p: i32 = 0;
        let expected: i32 = 0;

        // act
        let actual = Solution::minimize_max(nums, p);

        // assert
        assert_eq!(expected, actual);
    }
}
